_stbt/power.py

```diff
@@ -0,0 +1,188 @@
+    remotes = [
+        (r'none', _NoPdu),
+        (r'file:(?P<filename>[^:]+)', _FilePdu),
+        (r'(?P<model>pdu|ipp|aten|testfallback):(?P<hostname>[^: ]+)'
+         '(:(?P<outlet>[^: ]+))?', _ShellPdu),
+        (r'usbpower8800(:(?P<filename>[^:]+))?', _new_usb_power_8800_pro),
+    ]
+    for regex, factory in remotes:
+        m = re.match(regex, uri, re.VERBOSE | re.IGNORECASE)
+        if m:
+            return factory(**m.groupdict())
+    raise ConfigurationError('Invalid PDU URI: "%s"' % uri)
+
+
+class _NoPdu(object):
+    def set(self, _power):
+        raise RuntimeError("No Computer Controlable Power Outlet Configured")
+
+    def get(self):
+        # If we can't turn it off, it must be on
+        return True
+
+
+class _FilePdu(object):
+    """PDU useful for testing"""
+    def __init__(self, filename):
+        self.filename = filename
+
+    def set(self, power):
+        with open(self.filename, 'w') as f:
+            f.write(['0', '1'][power])
+
+    def get(self):
+        try:
+            with open(self.filename, 'r') as f:
+                return bool(int(f.read(1)))
+        except IOError as e:
+            if e.errno == errno.ENOENT:
+                return True
+            else:
+                raise
+
+
+class _ShellPdu(object):
+    """
+    stbt-power used to be written in bash, supporting three different types of
+    hardware.  This is a wrapper to allow the old bash script to continue
+    working until it can be removed entirely.
+    """
+    def __init__(self, model, hostname, outlet=None):
+        uri = '%s:%s' % (model, hostname)
+        if outlet is not None:
+            uri += ':' + outlet
+        self.cmd = ['bash', os.path.dirname(__file__) + "/stbt-power.sh",
+                    '--power-outlet=%s' % uri]
+
+    def set(self, power):
+        import subprocess
+        subprocess.check_call(self.cmd + [["off", "on"][power]])
+
+    def get(self):
+        import subprocess
+        power = subprocess.check_output(self.cmd + ["status"]).strip()
+        return {'ON': True, 'OFF': False}[power]
+
+
+class _UsbPower8800Pro(object):
+    """Documentation of the serial IO protocol found on the Aviosys website:
+
+    http://www.aviosys.com/downloads/manuals/power/USB%20Net%20Power%208800%20Pro%20Manual_EN.pdf
+
+    >>> f = _FakeUsbPower8800()
+    >>> u = _UsbPower8800Pro(f)
+    >>> u.get()
+    False
+    >>> u.set(True)
+    >>> f.is_on
+    True
+    >>> u.get()
+    True
+    >>> u.set(False)
+    >>> f.is_on
+    False
+    >>> u.get()
+    False
+    """
+    def __init__(self, device):
+        """Device is a file-like serial device"""
+        self.device = device
+
+    def set(self, power):
+        self.device.write("p1=%i\n" % power)
+        self.device.readline()
+
+    def get(self):
+        self.device.write("readio\n")
+        self.device.readline()
+        return {
+            'IO:5\r\n': True,
+            'IO:0\r\n': False,
+        }[self.device.readline()]
```

Comment:
If this was written in a less condensed manner you'd get better error messages, i.e.

```
response = self.device.readline()
if response == "IO:5\r\n":
    return True
elif response == "IO:0\r\n":
    return False
else:
    raise RuntimeError("Unexpected response from Power 8800 device: %s" % line.strip())
```